Azure parse_network_config uses fallback cfg when generate IMDS network cfg fails

johnsonshi opened the pull request:
Azure datasource's `parse_network_config` throws a fatal uncaught exception when an exception is raised during generation of network config from IMDS metadata. This happens when IMDS metadata is invalid/corrupted (such as when it is missing network or interface metadata). This causes the rest of provisioning to fail.

This changes `parse_network_config` to be a non-fatal implementation. Additionally, when generating network config from IMDS metadata fails, fall back on generating fallback network config (`_generate_network_config_from_fallback_config`).

This also changes fallback network config generation (`_generate_network_config_from_fallback_config`) to blacklist an additional driver: `mlx5_core`.

johnsonshi commented:
@anhvoms @Moustafa-Moustafa @trstringer This prevents invalid/corrupted IMDS network metadata from causing provisioning as a whole to fail.

anhvoms commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -1338,76 +1338,101 @@ def load_azure_ds_dir(source_dir):
+    """
+    netconfig = {'version': 2, 'ethernets': {}}
+    report_diagnostic_event(
+        'Azure: generating network configuration from IMDS')
```
I would drop the "Azure: " prefix, it's redundant and the rest of the code from this file doesn't use it. In fact I don't think you need this diagnostic event, it does not add any additional information. The telemetry reporter will generate telemetry to indicate that the code is invoking this function, which means it's generating configuration from IMDS

anhvoms commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -1338,76 +1338,101 @@ def load_azure_ds_dir(source_dir):
+    @return: Dictionary containing network version 2 standard configuration.
+    """
+    blacklist = ['mlx4_core']
+    report_diagnostic_event('Azure: generating fallback configuration')
```
Same as above.

anhvoms commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -1338,76 +1338,101 @@ def load_azure_ds_dir(source_dir):
-                blacklist_drivers=blacklist, config_driver=True)
-            evt.description = "network config from fallback"
-        return netconfig
+    if imds_metadata != sources.UNSET and imds_metadata:
+        try:
+            return _generate_network_config_from_imds_metadata(
+                imds_metadata)
+        except Exception as e:
+            msg = 'Failed generating network config from IMDS metadata: %s' % e
+            report_diagnostic_event(msg)
+            LOG.warning(msg)
+    try:
+        return _generate_network_config_from_fallback_config()
+    except Exception as e:
+        msg = 'Failed generating fallback network config: %s' % e
+        report_diagnostic_event(msg)
+        LOG.warning(msg)
+    return {}
```
Since you already have logging/telemetry independently for each of the if/else path, I think we can just consolidate this

```suggestion
    try:
        if imds_metadata != sources.UNSET and imds_metadata:
            return _generate_network_config_from_imds_metadata(imds_metadata)
        else:
            return _generate_network_config_from_fallback_config()
    except Exception as e:
        msg = 'Failed generating network config: %s' % e
        report_diagnostic_event(msg)
        LOG.warning(msg)
        return {}
 
```

johnsonshi replied:
Removed the excessive verbose logging without causing `test_parse_network_config_uses_fallback_cfg_when_no_network_metadata` and `test_parse_network_config_uses_fallback_cfg_when_no_interface_metadata` UTs to fail. This ensures that when an exception is raised during IMDS network config generation, cloud-init will fall back and generate fallback net cfg.

johnsonshi commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -1338,76 +1338,93 @@ def load_azure_ds_dir(source_dir):
-                blacklist_drivers=blacklist, config_driver=True)
-            evt.description = "network config from fallback"
-        return netconfig
+    if imds_metadata != sources.UNSET and imds_metadata:
```
@anhvoms I think we still need either a LOG.warning or a LOG.error (not a report_diagnostic_event) to be logged here for the VM's users to know if network config generation fails.

johnsonshi commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -1338,76 +1338,93 @@ def load_azure_ds_dir(source_dir):
-                blacklist_drivers=blacklist, config_driver=True)
-            evt.description = "network config from fallback"
-        return netconfig
+    if imds_metadata != sources.UNSET and imds_metadata:
+        try:
+            return _generate_network_config_from_imds_metadata(imds_metadata)
+        except Exception:
+            pass
+    try:
+        return _generate_network_config_from_fallback_config()
+    except Exception:
+        pass
+    return {}
```
@anhvoms I think we still need either a `LOG.warning` or a `LOG.error` (not a `report_diagnostic_event`) to be logged here for the VM's users to know if network config generation fails.

johnsonshi replied:
Added `LOG.error`.

johnsonshi commented on tests/unittests/test_datasource/test_azure.py:
```diff
@@ -156,8 +156,19 @@ def construct_valid_ovf_env(data=None, pubkeys=None,
+    }
 
-    def test_single_ipv4_nic_configuration(self):
+    @mock.patch('cloudinit.sources.DataSourceAzure.device_driver',
```
@TheRealFalcon @blackboxsw, in #539, the PR added a netplan driver filter when using the `hv_netvsc` driver.

However, that PR causes these UTs to fail execution on machines where the networking device driver is `hv_netvsc`. I managed to catch this as my machine's networking device driver happens to be `hv_netvsc`. This is due to the `device_driver` not being mocked out in several UTs. I'll be including these required `device_driver` mocks in this UT instead.

blackboxsw replied:
Good test fixes thanks.

blackboxsw commented:
@johnsonshi thanks for this PR, could you attach as a comment the response of cloud-init query -all (I'm specifically interested in the network section as surfaced by IMDS on one of these failed nodes). If reproducible

blackboxsw commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -1338,76 +1338,94 @@ def load_azure_ds_dir(source_dir):
+    if imds_metadata != sources.UNSET and imds_metadata:
+        try:
+            return _generate_network_config_from_imds_metadata(imds_metadata)
+        except Exception:
+            LOG.error(
+                'Failed generating network config from IMDS network metadata')
```
Would it be worth logging the str(eexception string so we have a hint of what went wrong?

```suggestion
        except Exception as e:
            LOG.error(
                'Failed generating network config from IMDS network metadata: %s', str(e))
```

johnsonshi replied:
Done

blackboxsw commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -1338,76 +1338,94 @@ def load_azure_ds_dir(source_dir):
+
+    @return: Dictionary containing network version 2 standard configuration.
+    """
+    blacklist = ['mlx4_core']
```
one note as well that we discussed with @rharper is that if we are generating fallback network config, we probably also need to blacklist "mlx5_core" drivers otherwise we'll get into a pickle on new hardware platforms with SRIOV devices too in fallback config we generate. So, maybe it's worth adding 'mlx5_core' to this fallback list too especially if we are driving more vms to fallthrough to fallback_config due to IMDS inconsistencies.

johnsonshi replied:
Done

johnsonshi commented on tests/unittests/test_datasource/test_azure.py:
```diff
@@ -1285,9 +1344,10 @@ def test_fallback_network_config_blacklist(self, mock_fallback, mock_dd,
 
         netconfig = dsrc.network_config
-        expected_config = fallback_config
-        expected_config['config'].append(blacklist_config)
```
@blackboxsw Is `test_fallback_network_config_blacklist` still a necessary UT? I discovered that this UT previously never failed because `expected_config = fallback_config` was a shallow copy rather than a deep copy.
Additionally, the UT `test_imds_network_config` already tests that `generate_fallback_config` is called with blacklist drivers in the arguments.
Additionally, the UT `test_device_driver_blacklist` in `tests/unittests/test_net.py` already tests the functionality of the device driver blacklist.
Since this UT (`test_fallback_network_config_blacklist`) does not accomplish much, I am of the opinion that it should be removed. Thoughts?

blackboxsw replied:
@johnsonshi Ugh, you are right on the shallow copy and I agree with you that we can drop `test_fallback_network_config_blacklist`  specifically because generate_fallback_config is actually tested in test_net.py as you mentioned. No need to integration test it from this side.

johnsonshi commented:
> One general question I have is if IMDS is in a state that cloud-init can't parse is IMDS network config content recoverable? As in, would retries buy us anything in these cases?

I'm afraid not. The IMDS inconsistencies and delays aren't deterministic, and it's impossible to "determine" whether the complete network metadata returned is even complete or not. The root cause of these inconsistencies is due to a platform issue.

johnsonshi commented:
> @johnsonshi thanks for this PR, could you attach as a comment the response of cloud-init query -all (I'm specifically interested in the network section as surfaced by IMDS on one of these failed nodes). If reproducible

We discovered these by looking at the `cloud-init.log` files of failed deployments. The IMDS response contained no `network` key in the instance metadata. We could wait until `network` is available in the response. However, normal behavior for IMDS is to *not* have `network` be delayed (it should be available way before). If there is a delay in network metadata availability, then it may be indicative of deeper platform issues (perhaps even the data within `network` isn't even complete).

```
2020-07-30 19:20:16,722 - handlers.py[DEBUG]: start: azure-ds/parse_network_config: 
2020-07-30 19:20:16,722 - DataSourceAzure.py[DEBUG]: Azure: generating network configuration from IMDS
2020-07-30 19:20:16,722 - handlers.py[DEBUG]: finish: azure-ds/parse_network_config: FAIL: 
2020-07-30 19:20:16,722 - util.py[WARNING]: failed stage init-local
2020-07-30 19:20:16,730 - util.py[DEBUG]: failed stage init-local
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 653, in status_wrapper
    ret = functor(name, args)
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 362, in main_init
    init.apply_network_config(bring_up=bool(mode != sources.DSMODE_LOCAL))
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 678, in apply_network_config
    netcfg, src = self._find_networking_config()
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 643, in _find_networking_config
    if self.datasource and hasattr(self.datasource, 'network_config'):
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceAzure.py", line 799, in network_config
    self._network_config = parse_network_config(nc_src)
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceAzure.py", line 1332, in parse_network_config
    network_metadata = imds_metadata['network']
KeyError: 'network'
```

blackboxsw commented on cloudinit/sources/DataSourceAzure.py:
```diff
@@ -1387,76 +1387,95 @@ def load_azure_ds_dir(source_dir):
+
+    @return: Dictionary containing network version 2 standard configuration.
+    """
+    blacklist = ['mlx4_core', 'mlx5_core']
```
@johnsonshi Thank you very much for this update. I was just talking today about how we needed this blacklist_driver update in tip of master due to my SRU verification of #539. Can you add a comment a suggested known instance type that should use mlx5_core driver so it'll ease testing and verification for our next SRU? I thought it was any v3 type advanced networking, but I wasn't sure.

johnsonshi replied:
I've just checked with the internal teams that handle Azure VM Instance SKUs. They've said that the [Dv4 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/dv4-dsv4-series) and the [Ev4 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/ev4-esv4-series) always have `mlx5`. Older Dv-? and Ev-? series VMs (such as [Dv2 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/dv2-dsv2-series), [Dv3 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/dv3-dsv3-series), and [Ev3 series VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/ev3-esv3-series)) may have either `mlx5` or `mlx4` hardware (with the older series ones having a higher chance of older `mlx4` hardware). I'll update the comment above with suggested instance types.

johnsonshi replied:
Comment added. Thanks @blackboxsw!

blackboxsw reviewed (commented):
Thanks @johnsonshi sorry for the delay here on response. I think we can drop that UT you mentioned and please update the pull request description to make note of the additional mlx5_core driver blacklist functionality on fallback config. 

We'll be using the pull request description as your squashed merge commit for this PR when this lands.

johnsonshi commented:
Apologies I've got a couple of high priority items so this has been delayed.

johnsonshi commented:
@blackboxsw I've updated the PR description + added the comments on the VM instance SKUs with `mlx4` and `mlx5` hardware. Also dropped the redundant UT as previously discussed.

blackboxsw reviewed (approved):
Excellent @johnsonshi thanks for the additions. I've launched and upgraded to this version cloud-init and see correct network configuration emitted and network properly setup on non-infiniband eth0 device.

```
ubuntu@SRU-worked-azure:~$ grep Trace /var/log/cloud-init.log 
ubuntu@SRU-worked-azure:~$ cloud-init status --long
status: done
time: Thu, 24 Sep 2020 16:43:35 +0000
detail:
DataSourceAzure [seed=/var/lib/waagent]
ubuntu@SRU-worked-azure:~$ cat /etc/netplan/50-cloud-init.yaml 
# This file is generated from information provided by the datasource.  Changes
# to it will not persist across an instance reboot.  To disable cloud-init's
# network configuration capabilities, write a file
# /etc/cloud/cloud.cfg.d/99-disable-network-config.cfg with the following:
# network: {config: disabled}
network:
    ethernets:
        eth0:
            dhcp4: true
            dhcp4-overrides: &id001
                route-metric: 100
            dhcp6: true
            dhcp6-overrides: *id001
            match:
                driver: hv_netvsc
                macaddress: 00:0d:3a:e2:d9:0e
            set-name: eth0
        eth1:
            dhcp4: true
            dhcp4-overrides: &id002
                route-metric: 200
            dhcp6: true
            dhcp6-overrides: *id002
            match:
                driver: hv_netvsc
                macaddress: 00:0d:3a:e2:de:17
            set-name: eth1
    version: 2
ubuntu@SRU-worked-azure:~$ ip addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:0d:3a:e2:d9:0e brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.4/24 brd 10.0.0.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 ace:cab:deca:deed::4/128 scope global dynamic noprefixroute 
       valid_lft 17279947sec preferred_lft 8639947sec
    inet6 fe80::20d:3aff:fee2:d90e/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:0d:3a:e2:de:17 brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.5/24 brd 10.0.0.255 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::20d:3aff:fee2:de17/64 scope link 
       valid_lft forever preferred_lft forever
4: rename4: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc mq master eth0 state UP group default qlen 1000
    link/ether 00:0d:3a:e2:d9:0e brd ff:ff:ff:ff:ff:ff
5: rename5: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc mq master eth1 state UP group default qlen 1000
    link/ether 00:0d:3a:e2:de:17 brd ff:ff:ff:ff:ff:ff
```
